examples/agents/ppo-chess-llm.py

```diff
@@ -0,0 +1,572 @@
+
+    return actor_llm_policy, data_llm_policy, critic_llm_policy, tokenizer
+
+
+def play(env, data_llm_policy, actor_llm_policy, tokenizer):
+    from torchrl.data import LazyStackStorage, ReplayBuffer
+
+    rb = ReplayBuffer(
+        storage=LazyStackStorage(100),
+        batch_size=48,
+        sampler=SliceSamplerWithoutReplacement(slice_len=8, end_key=("next", "done")),
+    )
+
+    # def breakpointy(td):
+    #     breakpoint()
+    #     return td
+
+    # rb.append_transform(breakpointy)
+
+    # Temporarily patched fbcode/pytorch/tensordict/tensordict/_lazy.py?lines=1502
+    rb.append_transform(lambda td: td.densify(layout=torch.jagged))
+
+    # rb.append_transform(breakpointy)
+
+    # obs_tokens in layout=torch.jagged errors with Qwen
+    # File "/home/mg1998/.conda/envs/rl/lib/python3.10/site-packages/transformers/models/qwen2/modeling_qwen2.py", line 859, in forward
+    # cache_position = torch.arange(
+    #             past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
+    #         )
+    # AttributeError: 'ConstantIntNode' object has no attribute 'add'
+    # After attempting to fix this there were other issues like NJT can't unsqueeze dim=0
+    def obs_token_transform(td):
+        obs_token_lengths = td["obs_tokens"].offsets().diff()
+        next_obs_token_lengths = td["next", "obs_tokens"].offsets().diff()
+        obs_token = td["obs_tokens"].to_padded_tensor(tokenizer.pad_token_id)
+        next_obs_token = td["next", "obs_tokens"].to_padded_tensor(
+            tokenizer.pad_token_id
+        )
+        # necessary because of https://github.com/pytorch/rl/blob/4c06ce2b81aa8871ac9e41478668843c5d48a39c/torchrl/objectives/value/advantages.py#L472-L477
+        obs_token, next_obs_token = pad_tensors_to_same_last_dim(
+            obs_token, next_obs_token, tokenizer.pad_token_id
+        )
+        td["obs_tokens"] = obs_token
+        td["next", "obs_tokens"] = next_obs_token
+        td["obs_token_lengths"] = obs_token_lengths
+        td["next", "obs_token_lengths"] = next_obs_token_lengths
+        return td
+
+    rb.append_transform(obs_token_transform)
+
+    collector = SyncDataCollector(
+        env, data_llm_policy, frames_per_batch=20, total_frames=10000
+    )
+    loss_module = ClipPPOLoss(
+        actor_network=actor_llm_policy,
+        critic_network=critic_llm_policy,
+    )
+
+    # FIXME: is this the right way to do this?
+    loss_module.tensor_keys.action = "tokenized_action"
+
+    optim = torch.optim.Adam(loss_module.parameters(), lr=2.5e-4)
+
+    gae = GAE(
+        value_network=Seq(actor_llm_policy[0], *critic_llm_policy),
+        gamma=0.99,
+        lmbda=0.95,
+        shifted=True,
+    )
+
+    for data in tqdm(collector):
+        # FIXME: reward seems to be getting wrongly propagated (e.g. sunfish's win gets reflected as llm's win)
+        rb.empty()
+        rb.extend(data)
+
+        for data in tqdm(rb):
+
+            data = gae(data)
```

Comment:
gae should go above the tqdm(rb)

You need the data to be presented sequentially to apply gae - even slices may not be ideal (better to compute it once and for all than at each iter).

Also I would put a no_grad around it for safety.

Reply:
this was something I wanted to talk about! if gae is before putting into replay buffer, we have no way to nicely append_transform to collector (I think)